pinecone_rag/ingestion.py

```
@@ -389,98 +704,181 @@ def _upsert_batch(
+                "Sparse upsert failed for batch %s, rolling back dense upsert: %s",
+                batch_num,
+                e,
+            )
+            if record_ids and namespace:
+                try:
+                    self.dense_index.delete(ids=record_ids, namespace=namespace)
+                    logger.info(
+                        "Rolled back %s records from dense index after sparse failure",
+                        len(record_ids),
+                    )
+                except Exception as rollback_e:
```

Ruff (0.15.5)

[warning] Do not catch blind exception: `Exception`

(BLE001)